libraries/AP_Eigen/AP_Eigen.h

```diff
@@ -0,0 +1,479 @@
+            // just a warning that the rounding is useless
+            if(!std::is_integral<IntType>::value || !std::is_floating_point<FloatType>::value) {
+                printf("VectorE2::get_nearest_integral(): template parameters do not make sense\n");
+                //std::cerr << "VectorE2::get_nearest_integral(): template parameters do not make sense" << std::endl;
+            }
+            
+            IntType vx = safeFloatToInt<IntType>(round(other[0]));
+            IntType vy = safeFloatToInt<IntType>(round(other[1]));
+            return VectorE2<IntType>(vx, vy);
+        }
+        /*
+         * @brief If <T> is a floating point type, 
+         * this functions rounds always down and returns it as <T>.
+         */
+        template <class IntType, class FloatType>
+        static VectorE2<IntType> get_integral(const VectorE2<FloatType> &other) {
+            // just a warning that the rounding is useless
+            if(!std::is_integral<IntType>::value || !std::is_floating_point<FloatType>::value) {
+                printf("VectorE2::get_integral(): template parameters do not make sense\n");
+                //std::cerr << "VectorE2::get_integral(): template parameters do not make sense" << std::endl;
+            }
+            
+            IntType vx = safeFloatToInt<IntType>(other[0]);
+            IntType vy = safeFloatToInt<IntType>(other[1]);
+            return VectorE2<IntType>(vx, vy);
+        }
+        
+        /*
+         * @brief constrains the values of the matrix to fit within a given range: low and high
+         */
+        void constrain(const T &low, const T &high) {
+            (*this)[0] = constrain_value<T>((*this)[0], low, high);
+            (*this)[1] = constrain_value<T>((*this)[1], low, high);
+        }
+        
+        /*
+         * @brief if the matrix contains angles (euler, radian), they can be constrained to fit between 180° and 360°
+         */
+        void constrain(const int flag) {
+            (*this)[0] = constrain_angle<T>((*this)[0], flag);
+            (*this)[1] = constrain_angle<T>((*this)[1], flag);
+        }
+        
+        /*
+         * @brief constrains the values of the matrix to fit within a given range: low and high
+         */
+        VectorE2<T> constrained(const T &low, const T &high) const {
+            VectorE2<T> tmp;
+            tmp[0] = constrain_value<T>((*this)[0], low, high);
+            tmp[1] = constrain_value<T>((*this)[1], low, high);
+            return tmp;
+        }
+        /*
+         * @brief if the matrix contains angles (euler, radian), they can be constrained to fit between 180° and 360°
+         */
+        VectorE2<T>  constrained(const int flag) const {
+            VectorE2<T> tmp;
+            tmp[0] = constrain_angle<T>((*this)[0], flag);
+            tmp[1] = constrain_angle<T>((*this)[1], flag);
+            return tmp;
+        }
+        
+        // some functions
+        float length() const {
```

Comment:
Eigen has a `norm()` family of functions that provide this. Maybe we should transition to use that?